packages/bloc/lib/src/bloc.dart

```diff
@@ -1,8 +1,149 @@
+  Future<void> onEach<T>(Stream<T> stream, void Function(T) onData);
+
+  // Subscribes to the provided [stream] and invokes the [onData] callback
+  /// when the [stream] emits new data and the result of [onData] is emitted.
+  ///
+  /// [forEach] completes when the event handler is cancelled or when
+  /// the provided [stream] has ended.
+  Future<void> forEach<T>(
+    Stream<T> stream,
+    FutureOr<State> Function(T) onData,
+  );
+
+  /// Whether the [EventHandler] associated with this [Emitter]
+  /// has completed.
+  /// [isCompleted] will be true either when the [EventHandler] has
+  /// completed normally or if it has been cancelled.
+  bool get isCompleted;
+
+  /// Emits the provided [state].
+  void call(State state);
+}
+
+/// An event handler is responsible for reacting to an incoming [Event]
+/// and can emit zero or more states via the [Emitter].
+typedef EventHandler<Event, State> = FutureOr<void> Function(
+  Event,
+  Emitter<State>,
+);
+
+/// Signature for a function which converts an incoming event
+/// into an outbound stream of events.
+/// Used when defining custom [EventTransformer]s.
+typedef EventMapper<Event> = Stream<Event> Function(Event);
+
+/// Used to change how events are processed.
+/// By default events are processed concurrently.
+///
+/// See also:
+///
+/// * [concurrent]
+/// * [drop]
+/// * [enqueue]
+/// * [restartable]
+typedef EventTransformer<Event> = Stream<Event> Function(
+  Stream<Event>,
+  EventMapper<Event>,
+);
+
+/// Process events concurrently. This is the default behavior.
+///
+/// **Note**: there may be event handler overlap and state changes will occur
+/// as soon as they are emitted. This means that states may be emitted in
+/// an order that does not match the order in which the corresponding events
+/// were added.
+EventTransformer<Event> concurrent<Event>() {
+  return (events, mapper) => events.flatMap(mapper);
+}
+
+/// Process events one at a time by maintaining a queue of added events
+/// and processing the events sequentially.
+///
+/// **Note**: there is no event handler overlap and every event is guaranteed
+/// to be handled in the order it was received.
+EventTransformer<Event> enqueue<Event>() {
+  return (events, mapper) => events.asyncExpand(mapper);
+}
+
+/// Process only one event by cancelling any pending events and
+/// processing the new event immediately.
+///
+/// Avoid using [restartable] if you expect an event to have
+/// immediate results -- it should only be used with asynchronous APIs.
+///
+/// **Note**: there is no event handler overlap and any currently running tasks
+/// will be aborted if a new event is added before a prior one completes.
+EventTransformer<Event> restartable<Event>() {
+  return (events, mapper) => events.switchMap(mapper);
+}
+
+/// Process only one event and drop any new events
+/// until the current event is done.
+///
+/// **Note**: dropped events never trigger the event handler.
+EventTransformer<Event> drop<Event>() {
```

Comment:
That's a good point -- currently the naming aligns with the convention in ember concurrency since that was the inspiration for a lot of these but I will raise this inconsistency and potentially see if we want to change any of the names entirely (for example maybe `sequential` rather than `enqueue`)?

Reply:
These operators have been implemented dozens of times probably by different stream frameworks, Rx impls, etc., and noone can seem to agree on naming. Instead of making up our own I'd personally just keep it consistent with at least one other thing that someone has chosen to name them before XD

Reply:
Haha yeah I agree it'd be nice to align with whatever is the most common/widely accepted naming convention. Imo `concurrent` and `sequential` are nice because they are both descriptive and easy to understand. I personally would prefer keeping those as is but I'm not sure about `drop` and `restartable`. The reason I suggested `switchLatest` is because that's the name of `switchMap`'s underlying API (after the mapper is applied). I don't really know what to do with `exhaustMap`/`drop` though 🤷

Reply:
Those seem fine to me; I still expect the common user may have no idea what they mean though (particularly the last two), so we'll definitely have to explain it well in the doc comments 😂

Reply:
@cmc5788 yeah agreed we'll definitely document it very thoroughly regardless of the naming.

@narcodico @jorgecoca @orestesgaolin what do you think?

Reply:
I like those last 4 suggestions, unfortunately `switch` is a reserved keyword.

Reply:
ah right...duh 🤦 I guess it's back to the drawing board 😛

Reply:
@felangel Maybe `concurrent` `sequential` `restartable` and `exhaustive` if we want to be grammatically consistent and name things to avoid keywords?

Reply:
That sounds pretty good to me 👍

Reply:
Since we're using `restartable` I went with `droppable` for now to align a bit more with ember concurrency's naming. I also experimented with googling "exhaustive [stream | concurrency | rx]" and it didn't really yield any useful results whereas "droppable concurrency" does point to ember concurrency docs 🤷. I'll try to get some more feedback tomorrow and hopefully just pick one and stick with it.

Reply:
I don't wanna ruin your grammar party folks, but I feel `restartable` and `droppable` are rarely ever gonna be used by themselves. I feel they almost always get used in conjunction with debounce/throttle/distinct/distinctUnique/etc., so is it even worth it supporting them out of the box ? 
That is unless we support chaining or something like `(events, mapper) => restartable((e) => e.debounce(...).distinct())`, but at this point you could easily just use `rxdart` or `stream_transform` and have transformers such as `(events, mapper) => events.debounce(...).distinct().switchMap(mapper)` and you're pretty much there while also reading a bit better.

Even `concurrent`/`sequential` bring limited functionality to the table. I see them useful more in terms of migrating to the new API with a specific default transformer, but even then they don't allow for full flexibility, they're more around the lines of specific use case convenience utils. Realistically you can't and shouldn't support all possible transformers cause then this would become rxbloc.

I'd suggest dropping them all or if you prefer making them `droppable` 😁 and keep the concurrent implementation internal as a default and then the users could make use of any external package they want to define their transformers. Alternatively, these transformers could be part of a separate package.

Reply:
Yeah I've gone back and forth on this quite a bit and I would also prefer to remove them all from the core library (concurrent would still remain the default but it wouldn't be exposed publicly). Since the package makes it fairly easy to define your own custom transformers with `rxdart` or `stream_transform` I too am worried about it becoming `rxbloc` 😛 

An idea I've run by multiple teammates has been to create a `bloc_concurrency` package which mimics the ember concurrency APIs for those who want to use them without needing to pull in something like `rxdart`. @cmc5788 how would you feel about that? I know you expressed interest in keeping `restartable` and `droppable` as part of the core library but given they didn't ship with previous versions of bloc and would be pretty easy to add (either manually or via something like `bloc_concurrency`) I feel it's a fair compromise in order to keep the core dart library simpler and easier to maintain without bloating it with rx operators.

Reply:
Well, restartable, drop, and concurrent are all implemented by [`stream_transform`](https://pub.dev/packages/stream_transform), and if it were up to me, even for a default of concurrent I certainly would not implement it myself and would use a trusted implementation like that.

So if you are going to rely on a trusted implementation by a dart core maintainer, why not expose the two others as well?

Maybe others are different, but I also know I would _definitely_ use `restartable` and `drop` in any app of scale. I don't think it creeps into the scope of being "rxbloc" to add these core ones. (and obviously `asyncExpand` comes for free).

So... in my opinion, those are the core operators and I would add them 100% of the time, and it would be a mistake to exclude them (or reimplement any of them instead of using a trusted package under the hood).

Reply:
> Well, restartable, drop, and concurrent are all implemented by stream_transform, and if it were up to me, even for a default of concurrent I certainly would not implement it myself and would use a trusted implementation like that.

The current `concurrent` implementation is pretty straightforward and is a simplification of the `stream_transform` implementation since we don't have to worry about non-broadcast streams, pauses/resumes, etc. when it comes to the internal event stream.

> So if you are going to rely on a trusted implementation by a dart core maintainer, why not expose the two others as well?

It's a slipperly slope in my opinion because we open the door to feature requests for things like `debounce`, `throttle`, etc. with no clear boundaries (they are also part of `stream_transform`).

> Maybe you use operators differently than I do, but I pretty strongly disagree with this. restartable and drop are plenty useful on their own. I use them all the time for guarding against overlapping operations without any of debounce/throttle/distinct/distinctUnique etc. If anything, throttles and debounces are the less common cases. Making operations mutually exclusive by cancellation or ignoring intervening events is the typical use case.

Everyone uses operators differently because it varies drastically based on the use-case and type of app. Another data point is no one ever asked for built-in support for switchMap or exhaustMap since the library was created so it seems reasonable to assume that most developers would either continue to use the library as is or rely on their stream manipulation library of choice.

Reply:
From a maintenance standpoint if you feel comfortable implementing a simplified version of `concurrent` and maintaining it yourself, it seems fine to proceed without any extra operators. Everyone can keep their own opinions about what the useful operators are and avoid the debate here 😄 -- and if your goal isn't to be a batteries-included framework like `ember`, then it also makes sense to avoid setting reasonable defaults and let everyone figure it out on their own.

Reply:
My two cents here: have a separate package, and keep the ember naming; that has worked really well with `blocTest` and no one has complained (and given our background, testing seems a bit more core than these transformers) 

If you change your mind in the future, you can always deprecate a package and bring it back to core, but doing the reverse operation is much more complicated.

Reply:
@felangel The only other thought this brings to mind, is if we endorse the model that we're expecting people to bring their own implementations, we might want to make it easy to also globally replace the default `concurrent` as well.

If I'm bringing my own stream utility framework to plug in, I would expect to be able to provide my own framework's implementation of `concurrent` as the global default so it's consistently implemented with the other things my framework does 🤔

Reply:
> @felangel The only other thought this brings to mind, is if we endorse the model that we're expecting people to bring their own implementations, we might want to make it easy to also globally replace the default `concurrent` as well.
> 
> If I'm bringing my own stream utility framework to plug in, I would expect to be able to provide my own framework's implementation of `concurrent` as the global default so it's consistently implemented with the other things my framework does 🤔
> 
> Probably also not pollute the global namespace with `concurrent()` anymore.

Yeah agreed `concurrent` would not be part of the global namespace and I agree you should be able to override the default transformer (so that you could for example make `asyncExpand` the default for backward compatibility purposes).

Reply:
> My two cents here: have a separate package, and keep the ember naming; that has worked really well with blocTest and no one has complained (and given our background, testing seems a bit more core than these transformers)

Yeah, I'm not here to debate about what operators are the `one pure path` 😂 all I can tell you is that I need to have flatMap/exhaustMap/switchMap to build an app that makes sense, so I'll have to bring those in somehow. If they're not in `bloc` it would be nice to have them in a package somewhere that is easy to find and integrate.

Reply:
> Maybe you use operators differently than I do, but I pretty strongly disagree with this. `restartable` and `drop` are plenty useful on their own. 

You missed the point 😉
My argument is that they address very specific use cases and they lack in composability. If they so happen to solve most of your requirements that doesn't mean they'll do that for everyone.

I don't feel like bloc's responsibility should be to implement all kind of rx style manipulation on streams when this can easily be achieved with packages targeting this specifically. Especially since it's not realistically possible to offer everything needed out of the box.

Reply:
> You missed the point 😉
> My argument is that they address very specific use cases and they lack in composability. If they so happen to solve most of your requirements that doesn't mean they'll do that for everyone.
> 
> I don't feel like bloc's responsibility should be to implement all kind of rx style manipulation on streams when this can easily be achieved with packages targeting this specifically. Especially since it's not realistically possible to offer everything needed out of the box.

Well, using `ember-concurrency` as a model for the map operators we chose was never meant to solve 100% of use cases, but I haven't changed my opinion that the set of core mapping operators (even without composition with throttle/debounce/etc) accomplishes a lot.

The evidence I cited for this is still `ember-concurrency`; it's has been around since 2014 and is quite stable / has a lot of mileage.

If your goal is to accomplish `everything needed` of course you won't, but you can get a lot with a little. That said it's not my job to define what `bloc's responsibility` is in this matter, and I'm not going to make the case for a more batteries-included framework if that's not what people want, I just want to get the operators that are most used and useful as easily as I can.

Reply:
> My argument is that they address very specific use cases and they lack in composability.

Also not really sure what this means. Whether functionality is split into more or fewer packages doesn't change how you'd accomplish composability:

```dart
EventTransformer<Event> exhaustive<Event>() {
  return (events, mapper) => events.exhaustMap(mapper);
}
```

could be easily "composed" for people who want to do so:

```dart
EventTransformer<Event> throttledExhaustive<Event>(/*params*/) {
  return (events, mapper) => events.throttle(/*params*/).exhaustMap(mapper);
}
```

If the core mapping operation (exhaustMap, switchMap, etc) was in the base `bloc` package or not doesn't change this at all.....

Reply:
> could be easily "composed" for people who want to do so:
> 
> ```dart
> EventTransformer<Event> throttledExhaustive<Event>(/*params*/) {
>   return (events, mapper) => exhaustive().call(events.throttle(/*params*/), mapper);
> }
> ```

This is the recipe for rxbloc 😉 Also not scalable and it's actually proving my point of this not being composable in a clean fashion. Ideally you would want the transformer to be: `throttle(...).exhaustive()`. Otherwise you'd have to define all kind of weird custom transformers to cover all there is, which basically ends up in `bloc` reinventing `rxdart`.

I would also love to have everything I need straight from the core package but don't think is feasible.

Reply:
I could make it look nicer for you using some extensions; this is just an example showing the simplest way to accomplish it using the API in this branch right now. I believe your statement was the approach `lack in composability`, which is clearly untrue. I agree we could clean up the API for it 😄 but streams are inherently composable in Dart, `rx` doesn't add this.

Reply:
The whole point is not if we can improve the API, it is whether this should be part of the core library or not 😁

Reply:
Right, I agree. I think we settled that. I don't particularly care whether it's part of `core bloc` or not. In fact `ember-concurrency` itself is separate from `ember` (it's an addon)!

At this point I'm just continuing to make the case for the package that adds this functionality back 😄

Reply:
> Right, I agree. I think we settled that. I don't particularly care whether it's part of `core bloc` or not. In fact `ember-concurrency` itself is separate from `ember` (it's an addon)!
> 
> At this point I'm just continuing to make the case for the package that adds this functionality back 😄 -- and to @jorgecoca 's point, if someday someone decided to factor that package back into `core`, it wouldn't sacrifice composability as far as I know.

Yeah I definitely think there's value in having a package to add that functionality 😄 .
Thoughts between `package:bloc_concurrency` and `package:bloc_transformers`?

Reply:
`package:bloc_concurrency` represents a more realistic (narrow) scope IMO

Reply:
`package:bloc_concurrency` -- if you're planning on only supporting the existing 4 transformers;
`package:bloc_transformers` -- if you're planning on adding more transformers in the future, since obviously not all of them will have to do with concurrency.